org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ProcessTaskManager.java

```diff
@@ -46,159 +50,128 @@ public class ProcessTaskManager {
-				this.sleepCount = -1;
-				try {
-					wait(100);
-				} catch (InterruptedException ignore) {
-					// ignore
+				while (next instanceof CompilationUnitDeclaration cu) {
+					elements.add(cu);
+					// optionally read more elements if already available:
+					next = this.units.poll();
 				}
-				this.sleepCount = 0;
-				next = this.units[this.currentIndex];
-			} while (next == null);
-		}
-
-		this.units[this.currentIndex++] = null;
-		if (this.currentIndex >= this.size)
-			this.currentIndex = 0;
-		if (this.sleepCount >= 1 && ++this.sleepCount > 4) {
-			notify(); // wake up processing thread to add next unit but only after removing some elements first
-			yield = this.sleepCount > 8;
-		}
-	}
-	if (yield)
-		Thread.yield();
-	return next;
-}
-
-private void compile() {
-	int unitIndex = this.startingIndex;
-	synchronized (this) { // wait until processingTask is assigned
-		@SuppressWarnings("unused")
-		Future<?> p = this.processingTask;
-	}
-	boolean noAnnotations = this.compiler.annotationProcessorManager == null;
-	while (this.processingTask != null) {
-		this.unitToProcess = null;
-		int index = -1;
-		boolean cleanup = noAnnotations || this.compiler.shouldCleanup(unitIndex);
-		try {
-			synchronized (this) {
-				if (this.processingTask == null) return;
-
-				this.unitToProcess = this.compiler.getUnitToProcess(unitIndex);
-				if (this.unitToProcess == null) {
-					this.processingTask = null;
-					return;
+				if (!elements.isEmpty()) {
+					if (next != null) {
+						// defer any stop signal until all CompilationUnitDeclaration read
+						this.lastSignal = next;
+					}
+					return elements;
 				}
-				index = unitIndex++;
-				if (this.unitToProcess.compilationResult.hasBeenAccepted)
-					continue;
+			} catch (InterruptedException interrupt) {
+				throw new AbortCompilation(true/* silent */, new RuntimeException(interrupt));
+			}
+			if (next instanceof Error error) {
+				throw error;
+			}
+			if (next instanceof RuntimeException runtimeException) {
+				throw runtimeException;
+			}
+			if (next == STOP_SIGNAL) {
+				return Collections.emptyList();
 			}
+			throw new IllegalStateException("Received unexpected element to process: " + String.valueOf(next)); //$NON-NLS-1$
+		} while (true);
+	}
 
-			try {
-				this.compiler.reportProgress(Messages.bind(Messages.compilation_processing, new String(this.unitToProcess.getFileName())));
-				if (this.compiler.options.verbose)
-					this.compiler.out.println(
-						Messages.bind(Messages.compilation_process,
-						new String[] {
-							String.valueOf(index + 1),
-							String.valueOf(this.compiler.totalUnits),
-							new String(this.unitToProcess.getFileName())
-						}));
+	private void processing() {
+		try {
+			int unitIndex = this.startingIndex;
+			boolean noAnnotations = this.compiler.annotationProcessorManager == null;
+			while (this.processing) {
+				int index = unitIndex++;
+				boolean cleanup = noAnnotations || this.compiler.shouldCleanup(index);
```

Comment:
The `shouldCleanup(index)` is running here with `startingIndex + 1`, while in the original code it was `startingIndex`. The `shouldCleanup()` compares `index < this.annotationProcessorStartIndex`  and it seem that this would be a bug here.

Reply:
the ++ is a postfix, i.e. "index" has the unincremented unitIndex. "unitIndex" is not reused - only "index" is.